sdks/python/apache_beam/io/mongodbio.py

```diff
@@ -139,50 +143,77 @@ def __init__(self,
+      last_timestamp_plus_one_sec = (last_doc_id.generation_time +
+                                     datetime.timedelta(seconds=1))
+      stop_position = objectid.ObjectId.from_datetime(
+          last_timestamp_plus_one_sec)
 
-    # get an estimate on how many documents should be included in a split batch
-    desired_bundle_count = desired_bundle_size // self.avg_doc_size
+    desired_bundle_size_in_mb = desired_bundle_size // 1024 // 1024
+    split_keys = self._get_split_keys(desired_bundle_size_in_mb, start_position,
+                                      stop_position)
 
     bundle_start = start_position
-    while bundle_start < stop_position:
-      bundle_end = min(stop_position, bundle_start + desired_bundle_count)
-      yield iobase.SourceBundle(weight=bundle_end - bundle_start,
+    for split_key_id in split_keys:
+      bundle_end = min(stop_position, split_key_id)
+      if bundle_start is None and bundle_start < stop_position:
+        return
+      yield iobase.SourceBundle(weight=desired_bundle_size_in_mb,
                                 source=self,
                                 start_position=bundle_start,
                                 stop_position=bundle_end)
       bundle_start = bundle_end
+    # add range of last split_key to stop_position
+    if bundle_start < stop_position:
+      yield iobase.SourceBundle(weight=desired_bundle_size_in_mb,
+                                source=self,
+                                start_position=bundle_start,
+                                stop_position=stop_position)
 
   def get_range_tracker(self, start_position, stop_position):
     if start_position is None:
-      start_position = 0
+      epoch = datetime.datetime(1970, 1, 1)
+      start_position = objectid.ObjectId.from_datetime(epoch)
     if stop_position is None:
-      stop_position = self.doc_count
-    return OffsetRangeTracker(start_position, stop_position)
+      last_doc_id = self._get_last_document_id()
+      # add one sec to make sure the last document is not excluded
+      last_timestamp_plus_one_sec = (last_doc_id.generation_time +
+                                     datetime.timedelta(seconds=1))
+      stop_position = objectid.ObjectId.from_datetime(
+          last_timestamp_plus_one_sec)
+    return _ObjectIdRangeTracker(start_position, stop_position)
 
   def read(self, range_tracker):
     with MongoClient(self.uri, **self.spec) as client:
-      # docs is a MongoDB Cursor
-      docs = client[self.db][self.coll].find(
-          filter=self.filter, projection=self.projection
-      )[range_tracker.start_position():range_tracker.stop_position()]
-      for index in range(range_tracker.start_position(),
-                         range_tracker.stop_position()):
-        if not range_tracker.try_claim(index):
+      all_filters = self.filter
+      all_filters.update({
+          '_id': {
+              '$gte': range_tracker.start_position(),
```

Comment:
What if the original filter already has gte/lt conditions on _id? I think it's fine to prohibit it (at construction time); but you can also take it into account and infer the start/stop position in split() from it.

Reply:
Added the filter merge logic.